include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
-    
-    return retval;
-  }
-  template<class ObserverRayDirInfo, class IncomingRayDirInfo>
-  static AnisotropicMicrofacetCache createForReflection(
-    const surface_interactions::Anisotropic<ObserverRayDirInfo> interaction, 
-    const LightSample<IncomingRayDirInfo> _sample)
-  {
-    return createForReflection(interaction.getTangentSpaceV(),_sample.getTangentSpaceL(),_sample.VdotL);
-  }
-  // transmissive cases need to be checked if the path is valid before usage
-  static bool compute(
-    out AnisotropicMicrofacetCache retval,
-    const bool transmitted, const float3 V, const float3 L,
-    const float3 T, const float3 B, const float3 N,
-    const float NdotL, const float VdotL,
-    const float orientedEta, const float rcpOrientedEta, out float3 H
-  )
-  {
-    float3 H;
-    const bool valid = IsotropicMicrofacetCache::compute(retval,transmitted,V,L,N,NdotL,VdotL,orientedEta,rcpOrientedEta,H);
-    if (valid)
-    {
-      retval.TdotH = dot(T,H);
-      retval.BdotH = dot(B,H);
-    }
-    return valid;
-  }
-  template<class ObserverRayDirInfo, class IncomingRayDirInfo>
-  static bool compute(
-    out AnisotropicMicrofacetCache retval,
-    const surface_interactions::Anisotropic<ObserverRayDirInfo> interaction, 
-    const LightSample<IncomingRayDirInfo> _sample,
-    const float eta
-  )
-  {
-    float3 H;
-    const bool valid = IsotropicMicrofacetCache::compute(retval,interaction,_sample,eta,H);
-    if (valid)
-    {
-      retval.TdotH = dot(interaction.T,H);
-      retval.BdotH = dot(interaction.B,H);
-    }
-    return valid;
-  }
-
-  float TdotH;
-  float BdotH;
+    using this_t = SIsotropicMicrofacetCache<T>;
+    using scalar_type = T;
+    using vector3_type = vector<scalar_type, 3>;
+    using matrix3x3_type = matrix<scalar_type, 3, 3>;
+
+    using ray_dir_info_type = ray_dir_info::SBasic<scalar_type>;
+    using isotropic_type = surface_interactions::SIsotropic<ray_dir_info_type>;
+    using sample_type = SLightSample<ray_dir_info_type>;
+
+    // always valid because its specialized for the reflective case
+    static this_t createForReflection(const scalar_type NdotV, const scalar_type NdotL, const scalar_type VdotL, NBL_REF_ARG(scalar_type) LplusV_rcpLen)
+    {
+        LplusV_rcpLen = rsqrt<scalar_type>(2.0 + 2.0 * VdotL);
+
+        this_t retval;
+
+        retval.VdotH = LplusV_rcpLen * VdotL + LplusV_rcpLen;
+        retval.LdotH = retval.VdotH;
+        retval.NdotH = (NdotL + NdotV) * LplusV_rcpLen;
+        retval.NdotH2 = retval.NdotH * retval.NdotH;
+
+        return retval;
+    }
+    static this_t createForReflection(const scalar_type NdotV, const scalar_type NdotL, const scalar_type VdotL)
+    {
+        float dummy;
+        return createForReflection(NdotV, NdotL, VdotL, dummy);
+    }
+    template<class ObserverRayDirInfo, class IncomingRayDirInfo NBL_FUNC_REQUIRES(ray_dir_info::Basic<ObserverRayDirInfo> && ray_dir_info::Basic<IncomingRayDirInfo>)
```

Comment:
wrong concept on `IncomingRayDirInfo` ?  isn't it supposed to be a sample ?

Reply:
if you had two ray-dir infos, you would not have `VdotL` available, you'd have to work it out yourself from `observer.getV()` and `incoming.getV()`

Reply:
aaah you're constraining the template params to `surface_interactions::SIsotropic` and `SLightSample`